python/test/conftest.py

```diff
@@ -19,9 +19,17 @@ def device(request):
 def fresh_triton_cache():
     with tempfile.TemporaryDirectory() as tmpdir:
         from triton import knobs
+
+        old_jit_cache_hook = getattr(knobs.runtime, "jit_cache_hook", None)
+        old_post_compile_hook = getattr(knobs.runtime, "jit_post_compile_hook", None)
+
         with knobs.cache.scope():
             knobs.cache.dir = tmpdir
-            yield tmpdir
+            try:
+                yield tmpdir
+            finally:
+                knobs.runtime.jit_cache_hook = old_jit_cache_hook
+                knobs.runtime.jit_post_compile_hook = old_post_compile_hook
```

Comment:
I think the bug is here:
https://github.com/triton-lang/triton/blob/4d6d47d6f3ec27113e746b4b9ded393066ce3273/python/triton/knobs.py#L299-L303

Your new hook mechanism requires a deep-copy, as you're modifying `jit_cache_hook` in-place rather than re-assigning it.

Reply:
Actually nvm, you didn't change `jit_cache_hook`. Then why is the shallow copy of `__dict__` not enough?